pkg/controller/controllercmd/builder.go

```diff
@@ -238,13 +242,42 @@ func (b *ControllerBuilder) Run(ctx context.Context, config *unstructured.Unstru
 		return err
 	}
 
-	leaderElection.Callbacks.OnStartedLeading = func(ctx context.Context) {
-		if err := b.startFunc(ctx, controllerContext); err != nil {
-			klog.Fatal(err)
+	// 10s is the graceful termination time we give the controllers to finish their workers.
+	// when this time pass, we exit with non-zero code, killing all controller workers.
+	// NOTE: The pod must set the termination graceful time.
+	leaderElection.Callbacks.OnStartedLeading = b.getOnStartedLeadingFunc(ctx, controllerContext, 10*time.Second)
+
+	leaderelection.RunOrDie(ctx, leaderElection)
+	return nil
+}
+
+func (b ControllerBuilder) getOnStartedLeadingFunc(ctx context.Context, controllerContext *ControllerContext, gracefulTerminationDuration time.Duration) func(ctx context.Context) {
+	return func(context.Context) {
+		stopCh := make(chan struct{})
+		go func() {
+			defer close(stopCh)
+			if err := b.startFunc(ctx, controllerContext); err != nil {
+				b.nonZeroExitFn(fmt.Sprintf("controllers failed with error: %v", err))
+			}
+		}()
+
+		// when the stopCh is closed before the context is cancelled, it means the controllers
+		// died before we asked them to die.
```

Comment:
This is wrong. You are producing a race here during normal shutdown. The controllers are fast, you will run into the "prematurely" case altough the flow is totally fine.

Reply:
@sttts the controllers can only finish when the context is cancelled... if the function above exit before the context is cancelled, it means something caused the controller to terminate, without error...

I think even if the normal shutdown is fast, the context must be cancelled and therefore the <-ctx.Done() will take over the case will cause the stoppedChan check pass. (i swapped them + I added one more check into stopped branch. Will push shortly.

Reply:
there is no determinism for multiple contexts.